### ISSUE TITLE
Windows Terminal text glitches when using vim !

### ISSUE THREAD

**Issue by KadaGuetouache** [microsoft/terminal#6865]
# Environment

```none
Windows build number: [run `[Environment]::OSVersion` for powershell, or `ver` for cmd]
Windows Terminal version (if applicable):
Microsoft Windows [Version 10.0.19041.329]

Windows Terminal
Version: 1.0.1811.0
Any other software?

Using WSL2

Vim VIM - Vi IMproved 8.1 (2018 May 18, compiled Apr 15 2020 06:40:31)                         

# Steps to reproduce
Inside vim open curly brackets and hit tab.
<!-- A description of how to trigger this bug. -->

# Expected behavior
should work just fine.
<!-- A description of what you're expecting, possibly containing screenshots or reference material. -->

# Actual behavior
After some typing the text will be all around and becomes messing and hard to read.
opening window splits in vim and start jumping between then we clear the glitch just to start type again and mess start again.
<!-- What's actually happening? -->

**Comment by DHowett** [comment]
> Any other software? 
You didn't list any other software.

* What version of vim?
* Are you using WSL?
* Are you using vim inside WSL?
* Are you using vim over SSH?
* If you are using ssh, what version of ssh?

**Comment by cryptomasdar** [comment]
Yes i'm using Vim version 8.1 inside WSL 2 (Ubuntu 20.04)

**Comment by cryptomasdar** [comment]
Here is a sample of the text glitch on WT 
![Capture](https://user-images.githubusercontent.com/41374785/87222243-8e575e80-c372-11ea-9a07-c7d00f31cd2f.PNG)

**Comment by cryptomasdar** [comment]
> > Any other software?
> 
> You didn't list any other software.
> 
> * What version of vim?
> * Are you using WSL?
> * Are you using vim inside WSL?
> * Are you using vim over SSH?
> * If you are using ssh, what version of ssh?

yes i'm using Vim version 8.1 inside WSL 2 (Ubuntu 20.04)

**Comment by cryptomasdar** [comment]
After some windows updates it seems like the issue has gone so i will close this Issue.

**Comment by KadaGuetouache** [comment]
Problem still exits It's been months now and I haven't found a solution for this problem. The only think that I can do is close and open windows terminal app.

**Comment by zadjii-msft** [comment]
Are you using any sort of vim plugins? 

>  The only think that I can do is close and open windows terminal app.

That seems like a bit of an aggressive solution - does just closing an re-opening vim reset the buffer to something reasonable?

**Comment by bluddy** [comment]
I'm experiencing the same issues with both vim and neovim. Opening a split and scrolling up/down causes serious visual issues.

**Comment by kvnmlm** [comment]
Scrolling up and down is a killer.  Have to figure this out cause this kills using WSL completely.

**Comment by bluddy** [comment]
Switch to conemu and it'll be fine. It's a Windows terminal bug.

**Comment by zadjii-msft** [comment]
@kvnmlm we closed this since we couldn't get more feedback from OP, but if you like to help debug:


> * What version of vim?
> * Are you using vim over SSH?
> * If you are using ssh, what version of ssh?

Also:
* Does this repro without your vim profile? If it doesn't, could you share your `.vimrc`?

**Comment by bluddy** [comment]
Stepping in to answer: this happens both in vim and neovim, and doesn't require plugins, though it's more apparent with various plugins. The contents of the buffer all glitch out to various degrees. This is a showstopper from the perspective of my workflow, which is why I didn't bother sticking with Windows Terminal once I realized it was a terminal-specific issue and not a WSL2 one.

**Comment by zadjii-msft** [comment]
Well, we'd love to help figure that out! Curious that we haven't heard more reports of this - you'd think that a pretty major bug like that would have more reports if it was more widespread. 

Is there anything else interesting about your setup? What shell are you using? Maybe you're running `vim` in `tmux` or `screen`? Maybe you've got `TERM` set to something weird?

Is there a specific set of steps that repros this consistently (ideally from an initially empty `vim` buffer)?

**Comment by bluddy** [comment]
Hmm. Trying to make the bugs appear again, I'm not having success. I do use vim in zsh and tmux, but I switched away from Windows Terminal several Windows Update cycles ago, as well as updating my laptop video drivers to the latest from Intel. I'll switch back to Windows Terminal and see if it comes back and report here if it does.

**Comment by kvnmlm** [comment]
> * What version of vim?
  :version
  VIM - Vi IMproved 8.1 (2018 May 18, compiled Nov 08 2021 14:21:34)

> * Are you using vim over SSH?

No, WSL on my machine.  New to using WSL, cygwin user for years.

At first windows terminal looked nice but it tailed off pretty quickly.

I live in putty sshed into linux boxes daily using tmux most of the time.  Love putty, tolerate tmux.

> * If you are using ssh, what version of ssh?
n/a
> Does this repro without your vim profile? If it doesn't, could you share your .vimrc?

Trying at the moment and it is working normally.  Very strange cause last night everything I touched was hosed.

Will use it throughout the evening and report back when I break it.  I'll ditch my very simple .vimrc after it breaks again.

**Comment by kvnmlm** [comment]
It's happening again.  I scroll down to bottom then as I scroll up it starts creating these lines which look like the vim cursor line.  I'll add bottom of file in the term window a lot of weird tildes like a term problem of some sort.
![image](https://user-images.githubusercontent.com/10077207/148502599-fce1b550-9ac1-4e16-b813-0854d4586398.png)

**Comment by kvnmlm** [comment]
Further debug dropping .vimrc and then going after the cursor line.   This is the .vimrc line which kills it:

     set cursorline

Removing it seems to fix things.   Time will tell if that is the root but looks that way to me at moment.

**Comment by kvnmlm** [comment]
I'm afraid I've abandoned microsoft terminal also.  wsltty and putty at the moment work great.  terminal looks pretty but unusable.

**Comment by Sebazzz** [comment]
This happens in Nano (like the nano that ships with git for Windows) too. After editing a few lines that wrap, it doesn't seem to properly keep track of the layout. The experimental setting `experimental.rendering.forceFullRepaint` in settings.json does not seem to help.

**Comment by Aadv1k** [comment]
Sad that this occurs so often, it is a pain to work with vim/NeoVim with windows terminal, here is my [vimrc](https://github.com/Aadv1k/Dotvim/blob/7abf2c2204313bf4ec8417f3bee06416fc1ac741/init.vim) and I keep facing the same glitched text issue, where the text sometimes appears on the signcolumn, or sometimes gets misplaces when using splits.

**Comment by mhchen** [comment]
I'm glad I'm not the only one after seeing so many people praise Window Terminal.

![image](https://user-images.githubusercontent.com/1325549/158038461-ccbaf51f-2476-4510-89a5-2d3757e92563.png)

This is my terminal with tmux open after having closed neovim. I ran a bunch of git commands but there were a bunch of artifacts that remained from vim. When my terminal is in this state, even `clear` doesn't work (it jumps my prompt to the top of the terminal but doesn't clear any of the text or artifacts). When I connect to the same tmux session in other terminals like Hyper, the artifacts are gone.

Windows Terminal is my favorite balance of speed and aesthetics, but this issue makes it entirely unusuable for me for my vim/tmux workflow.

**Comment by mhchen** [comment]
![image](https://user-images.githubusercontent.com/1325549/158038688-b5abe5ac-c3b6-4bf3-ae0c-d9a64b1a97e1.png)


Here's another example of me opening up a popup/float window using [COC](https://github.com/neoclide/coc.nvim) in Neovim while using tmux. It renders totally incorrectly. Again, no problems with this using other terminals, even when connected to the same tmux session. Here's a screenshot of the same section in Hyper:

![image](https://user-images.githubusercontent.com/1325549/158038701-5294423f-a5b6-44f2-aea6-52be384ae7c9.png)

**Comment by Aadv1k** [comment]
![wt-glitch-2](https://user-images.githubusercontent.com/81357878/158050114-e13993f0-0c35-488f-af3f-e8538305766f.png)
![wt-glitch-1](https://user-images.githubusercontent.com/81357878/158050115-901281c0-8133-468d-be54-0d7b30ac5196.png)

The problem with these glitches is that they persist until I quit my entire vim session

**Comment by zadjii-msft** [comment]
~@Aadv1k & @mhchen you're likely running into #6987, or at least the artifact you're reporting looks more like that issue.~

...

Maybe, maybe not though. Let's be sure about that. 


@Aadv1k thanks for sharing your vimrc - could I get a couple other pieces of info, just to narrow things down?

* ~Are you using vim in WSL? In mingw/git bash / cygwin? (if so which version?)~ oh okay from the screenshot you're using WSL, that keeps that easy.
* Are you sing any sort of ssh? 
* Does this need tmux to repro? Are you using any sort of tmux config?
* which vim version specifically are you using?
* Is there an exact set of steps to repro the issue? Like, "open {this file}, then make a vertical split, then start scrolling"? The actual file contents may be relevant here...

Cause like, look at this snippet:
![image](https://user-images.githubusercontent.com/18356694/158358717-cbc9c838-ce1d-47ee-a9be-b4f374294dfb.png)

What's that red box char there? Maybe that's something that's ambiguous width, and what's confusing conpty here. The following line has the artifact on it, so maybe that triggered it.


<hr> I'm tempted to close this thread and punt back to  #6987. OP is MIA so I can't really get any more info to debug from them. The second artifact in this issue looks like #9359, but that poster has stopped using the Terminal so it's gonna be hard to get more info from them

**Comment by Aadv1k** [comment]
To answer your questions line by line;
> Are you sing any sort of ssh?

Nope, this is Ubuntu 20.04 running on wsl2 

> Does this need tmux to repro? Are you using any sort of tmux config?

No, it even happens without tmux, the main culprit seems to be vim's navigations and splits here which cause the tearing

> which vim version specifically are you using?

Sorry for not specifying but I am not using vim at all I am using `NVIM v0.6.1`

> Is there an exact set of steps to repro the issue? Like, "open {this file}, then make a vertical split, then start scrolling"? The actual file contents may be relevant here...

Open any file, and then just start navigating around, open vertical splits, horizontal splits, navigate across the file, that's pretty much it, there aren't any "steps" this just happens while navigating, even worse when using `set cursorline` 

> What's that red box char there? Maybe that's something that's ambiguous width, and what's confusing conpty here. The following line has the artifact on it, so maybe that triggered it.

That is actually my lsp, I use [nvim-lspconfig](https://github.com/neovim/nvim-lspconfig) tearing never usually happens around my lsp, such  issues happen around the signcolumn and/or the number column. 

I really hope this is resolved as it is really a pain to not be able to use such a great terminal like windows terminal.

**Comment by Aadv1k** [comment]
> > > Is there an exact set of steps to repro the issue? Like, "open {this file}, then make a vertical split, then start scrolling"? The actual file contents may be relevant here...
> > 
> > 
> > Open any file, and then just start navigating around, open vertical splits, horizontal splits, navigate across the file, that's pretty much it, there aren't any "steps" this just happens while navigating, even worse when using `set cursorline`
> 
> I'm aware that this is gonna sound _super_ annoying, but do you have a _specific_ set of steps that repros this 100%? I mean an exact "open _this_ .scss file, scroll to line 80, open a new vertical split, then open nerdtree with `:NERDTree`, Press `I`, {do whatever}"? I played with your vimrc for like half an hour this morning and couldn't get it to repro. Same with the folks in #6987 - while it does seem consistently to affect the people it affects, finding an exact repro has been really tricky. I really just want to get these two knocked out, since they're super breaking for the people they affect.

This is the strangest part about this, there really aren't any steps that can reproduce this exactly, it just randomly occurs, I try to document whenever it does happen, so I will update.

For the most part, having `set cursorline` and extensively using the visual mode in splits seem to be the biggest culprits in this case.

**Comment by Aadv1k** [comment]
> > > Is there an exact set of steps to repro the issue? Like, "open {this file}, then make a vertical split, then start scrolling"? The actual file contents may be relevant here...
> > 
> > 
> > Open any file, and then just start navigating around, open vertical splits, horizontal splits, navigate across the file, that's pretty much it, there aren't any "steps" this just happens while navigating, even worse when using `set cursorline`
> 
> I'm aware that this is gonna sound _super_ annoying, but do you have a _specific_ set of steps that repros this 100%? I mean an exact "open _this_ .scss file, scroll to line 80, open a new vertical split, then open nerdtree with `:NERDTree`, Press `I`, {do whatever}"? I played with your vimrc for like half an hour this morning and couldn't get it to repro. Same with the folks in #6987 - while it does seem consistently to affect the people it affects, finding an exact repro has been really tricky. I really just want to get these two knocked out, since they're super breaking for the people they affect.

I won't be writing any further comments here as the issue might not be windows terminal at all, I have been using it extensively and no issues, at least for now.

**Comment by EnGamma** [comment]
I propose reopening this issue.   I don't think this is a duplicate of #6987.   

I'm seeing screen corruption with VIM in Windows Terminal (WT) immediately upon startup of VIM and am NOT using tmux.   I've determined that it has something to do with VIM not getting the proper screen width from WT.   The most obvious indicator that this is happening is that the tail end of the VIM status bar get's wrapped to an added screen line.   For example, the tail end of the status bar might indicate the cursor is on line 141, column 1, 74% through the file with this  at the tail end: "141,1     74%".    What happens is "41,1 74%" gets wrapped to a new line on the screen.    Any scrolling causes more wrapping lines to appear at the bottom of the WT window.

I've found 2 cumbersome workarounds for now:

1.  Resize the terminal widow or
2.  <ALT>-<ENTER> <ALT>-<ENTER> to go to full screen and back

Performance is normal after that.

**Comment by Aadv1k** [comment]
> I propose reopening this issue. I don't think this is a duplicate of #6987.
> 
> I'm seeing screen corruption with VIM in Windows Terminal (WT) immediately upon startup of VIM and am NOT using tmux. I've determined that it has something to do with VIM not getting the proper screen width from WT. The most obvious indicator that this is happening is that the tail end of the VIM status bar get's wrapped to an added screen line. For example, the tail end of the status bar might indicate the cursor is on line 141, column 1, 74% through the file with this at the tail end: "141,1 74%". What happens is "41,1 74%" gets wrapped to a new line on the screen. Any scrolling causes more wrapping lines to appear at the bottom of the WT window.
> 
> I've found 2 cumbersome workarounds for now:
> 
> 1. Resize the terminal widow or
> 2. Toggle full screen [F11] [F11] or [ALT]-{ENTER] [ALT]-[ENTER]
> 
> Performance is normal after that.

Adding to this, for some reason even after setting padding to be 0, there is still a thin border at bottom and left of the windows terminal screen, doesn't go in full screen

**Comment by zadjii-msft** [comment]
@EnGamma That sounds like a different bug than this one - mind filing a new issue? 

<hr>

@Aadv1k That too is a different issue - in your case, the space that's there on the bottom of the window isn't big enough for an entire row of text, so we just fill it with blank space. That's generally preferable to a row of characters that's only partially visible. Otherwise, the only solution would be to prevent the window from filling the entire screen while maximized, which I'm sure that no one really wants. See also #6021

**Comment by mj-hd** [comment]
I'm facing the text glitches without tmux. (with neovim v0.7.2)

left: Windows Terminal 1.14.1963.0
right: WSL Terminal
https://user-images.githubusercontent.com/6854255/183268657-b9572ffb-63ab-4de3-ac41-db875d012bb4.mp4

When I move the cursor around the virtual texts, the texts below shift by one line.
This issue prevents me to use Windows Terminal for daily development...

Should I open another issue for this, or this issue should be reopened?

**Comment by EnGamma** [comment]
What you are seeing was also a symptom I experienced due to my trying to
set screen dimensions in my vimrc file.   The apparent cursor position was
different from actual which became evident when scrolling or typing which
would insert text in wrong place.

My particular problem was fixed by not trying to set screen dimensions in
vimrc.

**Comment by Aadv1k** [comment]
This issue seems to be on and off pretty regularly. From this thread and expereinces - this specific issue is caused in neovim configured with LSP or similar derivates that make use of signcolumns. The issue is ususally find alongside 
- init.vim with `set cursorline` opt
- Nerd fonts or other fonts that cause such artifacts (In my case, code new roman was causing issues) 
- Extensive navigation around splits and usage of visual mode
- Other funky settings that mess with the rendering of the screen (colorschemes, LSP messages, statuslines) 

NOTE: This is merely an observation from the long threads above, and includes some of my own experiences as well.

**Comment by zadjii-msft** [comment]
@mj-hd That looks like a different issue than the one tracked at the start of this thread. I'd file a new issue, and make sure to include your vimrc / any other relevant config.

**Comment by mj-hd** [comment]
Ok, thank you. I will check my vimrc and fonts to avoid these glitches, then open a new issue if the glitches still exist.

**Comment by bluebrown** [comment]
I also have glitches in neovim and helix editor. I am sure it's not a misconfiguration, but a Windows terminal issue.

**Comment by jradam** [comment]
Not sure if this is the same issue, but glitches start occurring for me only after running a terminal process within Vim. 

Once a process is running, the glitches slowly start occurring everywhere - and keep occurring even after quitting the process. 

The glitches take a while to creep in, but can be triggered instantly by opening a floating window:

https://user-images.githubusercontent.com/15231996/217013794-498cfaf4-6eda-4a58-aedb-95e6b1ae08eb.mp4
